inc-complete-derive/src/lib.rs

```diff
@@ -0,0 +1,678 @@
+///
+/// This generates a call to `define_input!(0, MyInput -> i32, MyStorage)`
+/// which provides the actual implementation.
+///
+/// Required attributes:
+/// - `id`: The unique computation ID (integer)
+/// - `output`: The output type
+/// - `storage`: The storage type
+///
+/// Optional attribute:
+/// - `assume_changed`: If present, indicates that the input is assumed to have changed
+#[proc_macro_derive(Input, attributes(inc_complete))]
+pub fn derive_input(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let struct_name = &input.ident;
+
+    // Find the inc_complete attribute
+    let inc_complete_attr = input
+        .attrs
+        .iter()
+        .find(|attr| attr.path().is_ident("inc_complete"));
+
+    let attr = match inc_complete_attr {
+        Some(attr) => attr,
+        None => {
+            return syn::Error::new(
+                Span::call_site(),
+                "Input derive requires #[inc_complete(...)] attribute",
+            )
+            .to_compile_error()
+            .into();
+        }
+    };
+
+    // Parse the attribute arguments using the Input-specific parser
+    let mut parser = InputAttributeParser::new();
+    if let Err(err) = parser.parse_attribute_list(attr) {
+        return err.to_compile_error().into();
+    }
+
+    let (id, output_type, storage_type, assume_changed) = match parser.validate(attr.span()) {
+        Ok(values) => values,
+        Err(err) => return err.to_compile_error().into(),
+    };
+
+    // Generate the define_input! call
+    let expanded = if assume_changed {
+        quote! {
+            inc_complete::define_input!(#id, assume_changed #struct_name -> #output_type, #storage_type);
+        }
+    } else {
+        quote! {
+            inc_complete::define_input!(#id, #struct_name -> #output_type, #storage_type);
+        }
+    };
+
+    TokenStream::from(expanded)
+}
+
+// =============================================================================
+// Intermediate Macros
+// =============================================================================
+
+/// Procedural macro for defining intermediate computations directly on functions
+///
+/// Usage:
+/// ```rust
+/// // Assumes ComputedDouble struct is already defined
+/// #[intermediate(id = 1)]
+/// fn compute_double(_context: &ComputedDouble, db: &DbHandle<MyStorage>) -> i32 {
+///     db.get(InputValue) * 2
+/// }
+/// ```
+///
+/// This generates a call to `define_intermediate!` using the extracted information.
+/// Assumes the computation struct is already defined.
+///
+/// Required attributes:
+/// - `id`: The computation ID (integer)
+///
+/// Optional attribute:
+/// - `assume_changed`: If present, indicates that the input is assumed to have changed
+///
+/// The macro automatically extracts:
+/// - **Output type** from the function return type (`i32`)
+/// - **Computation type** from the first parameter (`ComputedDouble` from `&ComputedDouble`)
```

Comment:
```suggestion
/// // Assumes ComputeDouble struct is already defined
/// #[intermediate(id = 1)]
/// fn compute_double(_context: &ComputeDouble, db: &DbHandle<MyStorage>) -> i32 {
///     db.get(InputValue) * 2
/// }
/// ```
///
/// This generates a call to `define_intermediate!` using the extracted information.
/// Assumes the computation struct is already defined.
///
/// Required attributes:
/// - `id`: The computation ID (integer)
///
/// Optional attribute:
/// - `assume_changed`: If present, indicates that the input is assumed to have changed
///
/// The macro automatically extracts:
/// - **Output type** from the function return type (`i32`)
/// - **Computation type** from the first parameter (`ComputeDouble` from `&ComputeDouble`)
```
Past tense -> present tense. "ComputedDouble" makes it sound like the computation has already happened, we want present tense so that when the user writes `ComputeDouble.get(db)` it sounds more like they're performing the computation. I suppose there's an argument to rename `get` to something like `call`, `perform`, or `run` though let's just be consistent for now with present tense.